Disable phone OTP option in select US area codes

jmhooper opened the pull request:
**Why**: Twilio does not allow us to call users in some US territories
which have a +1 internation code. This commit adds frontend code that
disables the phone option and messages the user that they will have to
receive an OTP via text.

jmhooper commented:
Right now this only includes code for the frontend part of this. I still need to figure how this looks on the backend.

jmhooper commented:
Initial screenshot:

![image](https://user-images.githubusercontent.com/963654/28379326-e3d74450-6c78-11e7-813c-49c2e17a6ae4.png)

zachmargolis commented on app/assets/javascripts/app/phone-internationalization.js:
```diff
@@ -0,0 +1,57 @@
+  const areaCode = areaCodeFromUSPhone(phone);
+  const country = PHONE_UNSUPPORTED_AREA_CODES[areaCode];
+  if (country) {
+    const messageTemplate = I18n.t('devise.two_factor_authentication.otp_delivery_preference.phone_unsupported');
```
might be worth expanding our custom `I18n.t` in JS to do templating?

ex

```js
I18n.t('devise.two_factor_authentication.otp_delivery_preference.phone_unsupported', { location: country })
```

zachmargolis commented on app/assets/javascripts/app/phone-internationalization.js:
```diff
@@ -0,0 +1,57 @@
+  const phoneRadio = document.querySelector('#two_factor_setup_form_otp_delivery_preference_voice');
+  const smsRadio = document.querySelector('#two_factor_setup_form_otp_delivery_preference_sms');
+  const phoneLabel = phoneRadio.parentNode.parentNode;
+  const deliveryMethodHint = document.querySelector('fieldset p');
```
did you mean to query the entire document? or just a part of a particular form?

jmhooper replied:
Ooops, meant to come back and clean this part up.

zachmargolis commented on app/assets/javascripts/app/phone-internationalization.js:
```diff
@@ -0,0 +1,57 @@
+const phoneFormatter = new PhoneFormatter();
+
+const PHONE_UNSUPPORTED_AREA_CODES = {
+  648: 'American Samoa',
```
Also I think we should move these values to something on the server, and expose them as `data-` attributes in the DOM or something (that way we can double check these values on the server which also works for clients without JS)

jmhooper replied:
Yeah, still working on the direction for how this looks on the server and thinking in that direction.

zachmargolis commented on spec/features/two_factor_authentication/sign_in_spec.rb:
```diff
@@ -32,6 +32,47 @@
     end
+
+    context 'with U.S. phone that does not support phone delivery method' do
+      let(:phone) { '671-555-5555' } # Guam phone number
```
what if we called the variable `guam_phone` and removed the comment?

jmhooper commented:
I've made some changes.

For forms with an OTP delivery method option, I added a validation to confirm that the OTP delivery method is supported for the given phone number. If it is not, it will add an error explaining the problem.

For forms without an OTP delivery method option (edit 2FA method form and idv phone confirmation) I added some code to make sure the OTP is sent via SMS and that the option to send a code via voice call is not presented to the user.

jmhooper commented:
Changed it up a little bit more so that the unsupported area codes are fed to the form in a data attribute. This should be ready for review now.

monfresh commented:
We should remove the "Mobile or landline okay" text, or make it dynamic such that it says "Mobile only" when voice is not supported.

monfresh commented on .rubocop.yml:
```diff
@@ -281,3 +281,8 @@ Style/TrailingCommaInLiteral:
+Layout/IndentHash:
+  Enabled: true
+  Exclude:
+    - '**/*.html.slim*'
```
Rubocop does not check Slim files. This can safely be removed.

jmhooper replied:
We have rubocop enabled for slim-lint. It runs rubocop on the files after slim processes them. I added a multiline hash to one and since slim removes all indentation this cop got angry.

There is a [`ignored_cops` options for slim-lint](https://github.com/sds/slim-lint/blob/master/lib/slim_lint/linter/README.md#rubocop), but adding one cop to that appears to re-enable all of the cops that slim-lint ignores by default.

monfresh replied:
Huh. This must be new. I didn't know you could disable cops within .rubocop.yml for slim-lint to ignore.

jmhooper commented on app/views/users/two_factor_authentication_setup/index.html.slim:
```diff
@@ -5,6 +5,9 @@ p.mt-tiny.mb0
     html: { autocomplete: 'off', role: 'form' },
+    data: {\
+      unsupported_area_codes: PhoneNumberCapabilities::VOICE_UNSUPPORTED_US_AREA_CODES.to_json,\
+    },
```
@monfresh: Here's the line that makes slim-lint upset. An alternative I'd prefer to disabling the `IndentHash` cop is figuring out how to make this a bit shorter.

monfresh replied:
Yes, I thought about that. You could set the hash in the controller and pass it to the view as a local variable.

zachmargolis commented on app/services/phone_number_capabilities.rb:
```diff
@@ -0,0 +1,32 @@
+  private
+
+  def area_code
+    @area_code ||= Phony.split(Phony.normalize(phone, cc: '1')).second
```
Personally I think using `Phonelib` here is clearer, like we do here? https://github.com/18F/identity-idp/blob/master/app/forms/otp_delivery_selection_form.rb#L37-L48

```ruby
Phonelib.parse(phone).area_code
```

jmhooper replied:
Fancy 👔. So many phone gems.....

jmhooper replied:
Alright, so this started breaking the spec for the IdV flow. The problem is that at this point the phone number has been formatted for the vendor and no longer includes the country code. It appears to still work for US numbers, but gets confused for numbers with a US country code.

Using Phony to normalize it to add the `+1` works, so I may do that.

Related: We may want to make sure we only accept US phone numbers for IdV in #1544

monfresh commented:
I'm seeing something weird in the UI when testing locally. See the Phone call button in the screenshot below:
<img width="564" alt="screen shot 2017-07-20 at 12 24 26 pm" src="https://user-images.githubusercontent.com/811150/28428169-7c66df6e-6d46-11e7-8830-9bbd99708d8f.png">

zachmargolis commented on app/services/phone_number_capabilities.rb:
```diff
@@ -0,0 +1,32 @@
+    true
+  end
+
+  def voice?
```
so it looks like in all the uses of this, we do `!capabilities.voice?` -- maybe should we just invert it and switch to `capabilities.sms_only?`

monfresh commented:
Here's another UI glitch on the OTP verification page:
<img width="472" alt="screen shot 2017-07-20 at 1 50 05 pm" src="https://user-images.githubusercontent.com/811150/28431472-763f8616-6d52-11e7-8a13-5da63c56c649.png">

monfresh commented:
One thing this PR doesn't address is that in the screen where you enter your new phone number when you want to change it, it doesn't mention anything about what kind of phone you can enter. Perhaps we should replace that view with the same view as in the Phone Setup?

jmhooper commented:
Re the UI issue on the OTP verification:

Looks like it also breaks if you have an authenticator app. It just says " or get a security code via authentication app."

I did also consider pulling the same UI we use for 2FA setup into the edit 2FA view, but wanted to pause on that and prioritize international calling and then come back to it if we have time at the end of this sprint or the beginning of the next one.

jmhooper commented:
Alright, I think I've addressed everything above^

monfresh commented on app/presenters/two_factor_auth_code/phone_delivery_presenter.rb:
```diff
@@ -42,7 +43,19 @@ def phone_number_tag
+
+    def otp_fallback_options_with_totp
+      if voice_otp_delivery_unsupported
+        safe_join([auth_app_fallback_tag, '.'])
```
Code Climate says this line is not tested.

jmhooper replied:
Ah yeah, this presenter has basically no tests. I can stick some in there.

jmhooper replied:
@monfresh: Added tests for the parts I changed.

monfresh commented on spec/support/shared_examples_for_otp_delivery_preference_validation.rb:
```diff
@@ -0,0 +1,24 @@
+  let(:phone) { '555-555-5000' }
+
+  before do
+    allow_any_instance_of(UpdateUser).to receive(:call)
```
We discourage the use of `allow_any_instance_of`, as do the RSpec docs: https://relishapp.com/rspec/rspec-mocks/docs/working-with-legacy-code/any-instance

The problem with this before statement is that it makes it look like `UpdateUser` should be called even if the validation failed, which is not true.

Instead, we do something like:
```ruby
update_user = instance_double(UpdateUser)
attributes = { otp_delivery_preference: 'voice' }
allow(UpdateUser).to receive(:new).with(user: user, attributes: attributes).and_return(update_user)
expect(update_user).to receive(:call) # when validation passes

capabilities = instance_double(PhoneNumberCapabilities)
allow(PhoneNumberCapabilities).to receive(:new).with(phone).and_return(capabilities)
allow(capabilities).to receive(:sms_only).and_return(false)

result = subject.submit(phone: phone, otp_delivery_preference: 'voice')
```
and
```ruby
expect(update_user).to_not receive(:call) # when validation fails
```

jmhooper replied:
@monfresh: Feels weird to stub new 🤔, but I went ahead and changed the test. I used a spy for the phone number capabilities because it handles some other calls to build the error message.

monfresh commented on spec/support/shared_examples_for_otp_delivery_preference_validation.rb:
```diff
@@ -1,20 +1,34 @@
+      attributes = { otp_delivery_preference: 'voice' }
+      allow(UpdateUser).to receive(:new).with(user: user, attributes: attributes).
+        and_return(update_user)
+      expect(update_user).to receive(:call) # when validation passes
```
Remove the comment?

monfresh commented on spec/support/shared_examples_for_otp_delivery_preference_validation.rb:
```diff
@@ -1,20 +1,34 @@
+      attributes = { otp_delivery_preference: 'voice' }
+      allow(UpdateUser).to receive(:new).with(user: user, attributes: attributes).
+        and_return(update_user)
+      expect(update_user).to_not receive(:call) # when validation passes
```
Remove the comment?

jmhooper replied:
That's what I get for copy/pasting 😳

jmhooper replied:
Done.
